File: src/leetCode/c++/Solution179.cpp
//O(n * logn * logZ) O(logn), n = nums.length, Z = num轉為字串的長度(最大32), 空間複雜度為排序所需
class Solution179 {
public:
    string largestNumber(vector<int>& nums) {
        sort(nums.begin(), nums.end(), [](int a, int b){//字串大排到小
            return to_string(a) + to_string(b) > to_string(b) + to_string(a);
        });
        string res = "";
        for(int num : nums)
            res += to_string(num);
        
        return res[0] == '0' ? "0" : res;//避免開頭為0的情況, ex:"00"
    }
};
/*nums=3,34,30

334 < 343 => 34 > 3
3430 > 3034 => 34 > 30
330 > 303 => 3 > 30

34 > 3 > 30
res=34330

*/

File: java/Solution179.java
package leetCode.java;

import java.util.*;

//O(n * logn * logZ) O(logn), n = nums.length, Z = num轉為字串的長度(最大32), 空間複雜度為排序所需
class Solution179 {
    public String largestNumber(int[] nums) {
        //Arrays.sort() lambda要放入包裝後的Integer或String, 先轉型
        String[] strs = new String[nums.length];
        for(int i=0; i<nums.length; i++)
            strs[i] = Integer.toString(nums[i]);

        Arrays.sort(strs, (a, b) -> (b + a).compareTo(a + b));//大排到小

        StringBuffer res = new StringBuffer();
        for(String str : strs)
            res.append(str);

        return res.charAt(0) == '0' ? "0" : res.toString();
    }
}